sim-cli/src/config.rs

```diff
@@ -0,0 +1,209 @@
+    pub print_batch_size: u32,
+    /// Level of verbosity of the messages displayed by the simulator.
+    /// Possible values: [off, error, warn, info, debug, trace]
+    pub log_level: LevelFilter,
+    /// Expected payment amount for the random activity generator
+    pub expected_pmt_amt: u64,
+    /// Multiplier of the overall network capacity used by the random activity generator
+    pub capacity_multiplier: f64,
+    /// Do not create an output file containing the simulations results
+    pub no_results: bool,
+    /// Duration after which results are logged
+    pub log_interval: u64,
+}
+
+impl<'de> Deserialize<'de> for SimulationConfig {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: serde::Deserializer<'de>,
+    {
+        #[derive(Deserialize)]
+        struct SimulationConfigHelper {
+            data_dir: PathBuf,
+            sim_file: PathBuf,
+            total_time: Option<u32>,
+            print_batch_size: u32,
+            log_level: String,
+            expected_pmt_amt: u64,
+            capacity_multiplier: f64,
+            no_results: bool,
+            log_interval: u64,
+        }
+
+        let helper: SimulationConfigHelper = Deserialize::deserialize(deserializer)?;
+
+        let log_level =
+            LevelFilter::from_str(&helper.log_level).map_err(serde::de::Error::custom)?;
+
+        Ok(SimulationConfig {
+            data_dir: helper.data_dir,
+            sim_file: helper.sim_file,
+            total_time: helper.total_time,
+            print_batch_size: helper.print_batch_size,
+            log_level,
+            expected_pmt_amt: helper.expected_pmt_amt,
+            capacity_multiplier: helper.capacity_multiplier,
+            no_results: helper.no_results,
+            log_interval: helper.log_interval,
+        })
+    }
+}
+
+impl SimulationConfig {
+    /// The default simulation configuration file with default simulation values
+    pub const DEFAULT_SIM_CONFIG_FILE: &'static str = "conf.json";
+
+    /// Loads the simulation configuration from a path and overwrites loaded values with
+    /// CLI arguments
+    pub fn load(
+        path: &PathBuf,
+        Cli {
+            data_dir,
+            sim_file,
+            total_time,
+            print_batch_size,
+            log_level,
+            expected_pmt_amt,
+            capacity_multiplier,
+            no_results,
+        }: &Cli,
+    ) -> anyhow::Result<Self> {
+        // 1. Load simulation config from specified path
+        let mut sim_conf = serde_json::from_str::<SimulationConfig>(
+            &std::fs::read_to_string(&path)
+                .context(format!("Failed to read content of {:?} to string", path))?,
+        )
+        .context("Failed to deserialize configuration string into SimulationConfig")?;
+
+        // 2. Override config values with CLI arguments (if passed)
+        if let Some(data_directory) = data_dir {
```

Comment:
This is quite a lot of code for similar functionality, can we do a macro to cut LOC?

Reply:
I do agree too. I have written a macro for this. However the log statements are never printed to the terminal because we [initialize](https://github.com/enigbe/sim-ln/blob/237475f7b3a98dbd53e7630be95a59e200602ff8/sim-cli/src/main.rs#L69C5-L80C19) the logger after loading and substituting configs.

```rust
macro_rules! cli_overwrite {
    ($config:expr, $cli:expr, $($field:ident),*) => {
        match (&mut $config, $cli) {
            (config_struct, cli_struct) => {
                $(
                    if let Some(value) = cli_struct.$field {
                        log::info!(
                            "SimulatinConfig::{} {:?} overwritten by CLI argument {:?}",
                            stringify!($field),
                            config_struct.$field,
                            value
                        );
                        config_struct.$field = value;
                    }
                )*
            }
        }

    };
}
```

I would like to replace the `simple_logger` to something that allows runtime update to log levels but not without agreement from others.

Reply:
Ah that's annoying, happy for you to shop around for a different logger 👍 

Thought we might be doing some more complex log-fu while we're running (which I didn't like the sound of), but if we're just going to set the level once off after we've parsed config works for me.